activerecord/lib/active_record/migration/command_recorder.rb

```diff
@@ -120,8 +120,12 @@ def #{method}(*args, &block)          # def create_table(*args, &block)
       alias :add_belongs_to :add_reference
       alias :remove_belongs_to :remove_reference
 
-      def change_table(table_name, **options) # :nodoc:
-        yield delegate.update_table_definition(table_name, self)
+      def change_table(table_name, **options, &block) # :nodoc:
+        if delegate.supports_bulk_alter? && options[:bulk]
+          @commands << [:bulk_change_table, [table_name, @reverting], block]
```

Comment:
I don't think we can defer running the block like this. There are admittedly limits to how much arbitrary ruby will work when we're running in reverse, but still this feels like its reordering would go a bit far.

Could we still do the recorder construction + yield in-line, and then enqueue the `apply_bulk_operations`?

Reply:
Hmmm, that sounds surmountable.

If necessary we could special-case it in `Migration#method_missing`, but before we do that... how about something like this?

```diff
--- a/activerecord/lib/active_record/migration/command_recorder.rb
+++ b/activerecord/lib/active_record/migration/command_recorder.rb
@@ -122,7 +122,8 @@ def change_table(table_name, **options) # :nodoc:
           recorder = self.class.new(self.delegate)
           recorder.reverting = @reverting
           yield recorder.delegate.update_table_definition(table_name, recorder)
-          @commands << [:bulk_change_table, [table_name, recorder.commands]]
+          commands = recorder.commands
+          @commands << [:change_table, [table_name], -> t { bulk_change_table(table_name, commands) }]
         else
           yield delegate.update_table_definition(table_name, self)
         end
```

(I don't have a suitable test case handy, so haven't verified the output it produces, but it sounds plausible?)

Reply:
Great idea @matthewd !

I applied your suggestion to our previous attempt. Running a reproduction code from the issue description now produces the following:

```ruby
D, [2022-07-08T11:42:03.551925 #17440] DEBUG -- :    (37.5ms)  DROP TABLE my_table
D, [2022-07-08T11:42:03.627596 #17440] DEBUG -- :    (75.4ms)  CREATE TABLE my_table (id INT)
==  BulkChange: migrating =====================================================
-- change_table(:my_table, {:bulk=>true})
D, [2022-07-08T11:42:03.652810 #17440] DEBUG -- :    (21.9ms)  ALTER TABLE `my_table` ADD `name` varchar(255), ADD `content` varchar(255)
   -> 0.0250s
==  BulkChange: migrated (0.0251s) ============================================

==  BulkChange: reverting =====================================================
-- change_table(:my_table)
D, [2022-07-08T11:42:03.676207 #17440] DEBUG -- :    (23.0ms)  ALTER TABLE `my_table` DROP COLUMN `name`, DROP COLUMN `content`
   -> 0.0232s
==  BulkChange: reverted (0.0233s) ============================================
```

The only downside is that `bulk => true` is not shown on the revert path. This could be fixed by passing the options in the recorder command but we'd have to use `Hash.ruby2_keywords_hash` to convert the hash to keyword arguments - not sure if it worth it.

Please let me know what you think.